File: DSA Using Java/Algorithms/Sorting Algorithms/MergeSort.java
public class MergeSort {
    public static void printArray(int[] arr){
        for (int element: arr){
            System.out.print(element+" ");
        }
        System.out.println();
    }
    public static void conquer(int[] arr, int startingIndex, int mid, int endingIndex){
        int[] mergedArray = new int[endingIndex-startingIndex+1];
        int index1 = startingIndex;
        int index2 = mid+1;
        int x = 0;
        while (index1 <= mid && index2 <= endingIndex){
            if (arr[index1] <= arr[index2]){
                mergedArray[x] = arr[index1];
                x++;
                index1++;
            }
            else {
                mergedArray[x] = arr[index2];
                x++;
                index2++;
            }
        }

        while (index1 <= mid){
            mergedArray[x] = arr[index1];
            x++;
            index1++;
        }

        while (index2 <= endingIndex){
            mergedArray[x] = arr[index2];
            x++;
            index2++;
        }

        // copying elements
        for (int i = 0, j = startingIndex; i < mergedArray.length; i++, j++){
            arr[j] = mergedArray[i];
        }
    }
    public static void divide(int[] arr, int startingIndex, int endingIndex){
        if (startingIndex >= endingIndex){
            return;
        }
        int mid = startingIndex + (endingIndex - startingIndex)/2;
        // int mid = (startingIndex+endingIndex)/2; // containing more space for larger values
        divide(arr, startingIndex, mid);
        divide(arr, mid+1, endingIndex);
        conquer(arr, startingIndex, mid, endingIndex);
    }
    public static void main(String[] args) {
        int[] arr = {7, 4, 9, 1, 2, 98, 0, 3, 4, 17, 97, 32, 11, 39};
        int n = arr.length;
        divide(arr, 0, n-1);
        printArray(arr);
    }
}

File: DSA Using C++/Algorithms/Sorting Algorithms/merge_sort.cpp
#include <iostream>

using namespace std;

// Note: Merge sort algorithm follows divide and conquer algorithm
// Time complexity = O(nlogn)
// Space complexity = O(n)

// Time: O(n)
void merge(int *arr, int st, int mid, int end)
{
    int i = st;
    int j = mid + 1;
    int n = end - st + 1; // size of array

    // auxilary array
    int *temp = new int[n];

    int index = 0;
    // arrange elements in temp array in sorted order
    while (i <= mid && j <= end)
    {
        if (arr[i] <= arr[j])
        {
            temp[index++] = arr[i++];
        }
        else
        {
            temp[index++] = arr[j++];
        }
    }

    while (i <= mid)
    {
        temp[index++] = arr[i++];
    }

    while (j <= end)
    {
        temp[index++] = arr[j++];
    }

    // copy sorted elements in original array
    for (int i = 0; i < n; i++)
    {
        arr[st + i] = temp[i];
    }

    // delete temp array
    delete[] temp; // free memory
}

// divide
// Time: O(logn)
void mergeSort(int *arr, int st, int end)
{
    if (st < end)
    {
        int mid = st + (end - st) / 2;
        mergeSort(arr, st, mid);
        mergeSort(arr, mid + 1, end);
        merge(arr, st, mid, end); // conqure
    }
}

// To print array
void printArray(int *arr, int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main()
{
    int arr[] = {3, 6, 7, 8, 9, 5, 4, 2, 1, 10};
    int n = sizeof(arr) / sizeof(int);

    mergeSort(arr, 0, n - 1);
    printArray(arr, n);

    return 0;
}